lib/ansible/plugins/vars/host_group_vars.py

```diff
@@ -55,33 +55,55 @@
 
 import os
 from ansible.errors import AnsibleParserError
-from ansible.module_utils._text import to_bytes, to_native, to_text
+from ansible.module_utils._text import to_bytes, to_native
 from ansible.plugins.vars import BaseVarsPlugin
 from ansible.inventory.host import Host
 from ansible.inventory.group import Group
+from ansible.utils.path import basedir
 from ansible.utils.vars import combine_vars
 
+CANONICAL_PATHS = {}
 FOUND = {}  # type: dict[str, list[str]]
+NAK = set()
+HOSTS = set()
+GROUPS = set()
```

Comment:
Maybe that would be simpler. Let me try that for comparison's sake at least.

Reply:
just so we have a few 3rd party to test against:
https://github.com/ansible-collections/community.sops/blob/main/plugins/vars/sops.py
https://github.com/manicminer/ansible-vars-plugin-aws/blob/master/vars_plugins/aws.py

tangent: this is just a simple example on writing vars plugins, several usable/testable ones. 
Aside from using for this ticket, I think we should incorporate a lot of this into our docs:
https://www.lilatomic.ca/posts/ansible_writing_facts_plugin/

Reply:
Even though I'm thinking now that variable manager would be a better place to do this in a generic way because of the other vars plugins, I'm not sure how to handle this for merge_hash behavior without worsening performance if there are a lot of host_vars/group_vars. To avoid accidentally mutating the cached variables, I think I'd need to deep copy the new variables before passing them to combine_vars. Keeping the cache at the plugin level allows loading the variables every time but caching some of the overhead that's unnecessary to do repeatedly.

Reply:
@bcoca I added a commit to try to cover some of the documentation gaps.